### ISSUE TITLE
tui: tmux with focus-events, slow "scan line" effect

### ISSUE THREAD

**Issue by nthapaliya** [neovim/neovim#7649]
- `nvim --version`: NVIM v0.2.2
- Vim (version: ) VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Jul 26 2017 19:10:24)
- Operating system/version:
- Terminal name/version: Iterm2 Build 3.1.5
- `$TERM`: screen-256color

Using the following vimrc (for both vim and nvim)
```
call plug#begin('~/.config/nvim/plugged')
Plug 'morhetz/gruvbox'
call plug#end()

set background=light " background light just to see the effect more quickly
colorscheme gruvbox
```

Using the following .tmux.conf
```
set -g focus-events on

set-option -ga terminal-overrides ",xterm-256color:Tc"
set-option -g default-terminal "screen-256color"

```
### Expected behaviour (Vim behavior)
Vim opens instantly

![vim](https://user-images.githubusercontent.com/4131655/33337418-853c84b4-d438-11e7-89e6-c17c688a1451.gif)

### Actual behaviour (nvim behavior)
Nvim opens but the screen takes a moment to "paint" from left to right, top to bottom

![nvim](https://user-images.githubusercontent.com/4131655/33337334-47a24e5e-d438-11e7-88b0-f43ca613c7e2.gif)

It seems it is `set -g focus-events on` in tmux that is causing this behavior. If I comment that line out and restart the tmux session, nvim behaves like vim.

**Comment by justinmk** [comment]
> set -g focus-events on

Hmm. I get the behavior on my local (macOS) tmux, but not remote (ubuntu) tmux. Both set `set -g focus-events on`. In the former case (macOS), there's no terminfo found for `screen-256color`, so Nvim's builtin terminfo definitions are used. I think the root cause lies there.

**Comment by nthapaliya** [comment]
`TERM=xterm-256color nvim -u temp-vimrc`

The issue still exists. A thought: @justinmk do the tmux versions on your remote ubuntu box and macOS machine match? I'm using tmux v2.6

**Comment by justinmk** [comment]
@nthapaliya the versions are different. The one that exhibits the problem is tmux 2.3, the one without the issue is 2.1. But that doesn't really matter, since Vim doesn't have the problem it is more likely a terminfo issue.

**Comment by nthapaliya** [comment]
Hmm, makes sense. Let me know if I can test something else.

**Comment by justinmk** [comment]
@nthapaliya If you build from source, you could try https://github.com/neovim/neovim/pull/7653 , then log the termcap info to a file, like this:

    ./build/bin/nvim -V3log

The `log` file will be flushed on exit, then please post it.

**Comment by nthapaliya** [comment]
@justinmk Ok, I did that. The logging didn't work as you described though: it would always crash at the end of the output and the log file would be empty. 

I did copy each page of logging manually to a text file. I'll post it here.

[nvim-log.txt](https://github.com/neovim/neovim/files/1512507/nvim-log.txt)

**Comment by justinmk** [comment]
This is indeed related to focus reporting. (I also compared terminfo files from the "good" case and "bad" case and found no significant differences)

Deferring this line in `tui.c` fixes this issue:

    unibi_out_ext(ui, data->unibi_ext.enable_focus_reporting);

**Comment by justinmk** [comment]
@nthapaliya Does https://github.com/neovim/neovim/pull/7664 fix the issue for you?

**Comment by nthapaliya** [comment]
@justinmk Yes! 

Well, mostly. Essentially, that patch fixes it. *However*, I can still see a difference between vanilla vim and neovim on pathological screen sizes (for example, by hitting cmd - multiple times on iterm and getting the font size to be *tiiny*)
```
~/P/neovim (tui-deferred-startup)> tput cols; tput lines
1270
264
```

In practice I never hit that case, so I would consider this fixed.

**Comment by justinmk** [comment]
> However, I can still see a difference between vanilla vim and neovim on pathological screen sizes (for example, by hitting cmd - multiple times on iterm and getting the font size to be tiiny)

@nthapaliya Does disabling tmux focus-events fix that case for Nvim?

**Comment by nthapaliya** [comment]
At those sizes, with tmux focus-events disabled, both vim and neovim appear to behave about the same. At those sizes vim flickers too though, sometimes. So it doesn't seem to be caused by whatever performance issue was specific to neovim + tmux-focus events

**Comment by dylan-chong** [comment]
For me, turning off `focus-events` in tmux removes the scanline effect, also removing from vimrc `colorscheme base16-materia` removes the scanline effect as well. Thought this might be use ful info

**Comment by dylan-chong** [comment]
I am using neovim 0.2.2_1, and can still replicate @nthapaliya 's bug. I assume @justinmk 's fix (#7664) was in this release version.

**Comment by justinmk** [comment]
> I assume @justinmk 's fix (#7664) was in this release version.

Why do you assume that? If you visit https://github.com/neovim/neovim/commit/27f9b1c7b029d8f61723fd261bed76ea7a01e68d , GitHub shows which tags/branches (nightly and master) include that commit.

**Comment by justinmk** [comment]
> I think it happens when the entire buffer needs to be redrawn (for example when doing an incsearch with /) AND when there are some other processes sucking up CPU power.

This issue was strictly about startup.

And yes, a full redraw can be slow. That's true for Vim, Nvim, and any other terminal application. A full redraw can be caused by plugins doing `:redraw!` or `:mode` or CTRL-L.

If you can bisect to a specific Nvim commit that caused a regression, or provide minimal steps to reproduce a case that you think shouldn't be causing a full redraw, report as a new issue.  But it's unrelated to this one.
